### PR TITLE
Miri: tweak error print

### PR DESCRIPTION
**Pull request by RalfJung**
I started by adjusting the "invalid use of int as pointer" message (it wasn't really clear what is invalid about the use). But then I realized that these are all `Debug` impls we use for these errors, for some reason, so I fixed that to use `Display` instead.

~~This includes https://github.com/rust-lang/rust/pull/71590 (to get the `Display` impl for `Pointer`), so the diff will look better once that finally lands. Here's the [relative diff](https://github.com/RalfJung/rust/compare/e72ebf5119e833b70231c3f2f8c7ca4904b1f0a3...RalfJung:miri-error-print).~~

r? @oli-obk

### CODE REVIEW

File: src/librustc_middle/mir/interpret/error.rs [hunk 1]
```diff
@@ -356,6 +354,8 @@ pub enum UndefinedBehaviorInfo {
     InvalidBool(u8),
     /// Using a non-character `u32` as character.
     InvalidChar(u32),
+    /// An enum discriminant was set to a value which was outside the range of valid values.
+    InvalidDiscriminant(ScalarMaybeUndef),
```
Comment (oli-obk) [inline src/librustc_middle/mir/interpret/error.rs:358]: why was this moved?

Comment (RalfJung) [reply src/librustc_middle/mir/interpret/error.rs:358]: I moved it to the other `Invalid*` variants. Invalid Boolean, invalid character, invalid discriminant -- all seem pretty related to me.

Comment (RalfJung) [reply src/librustc_middle/mir/interpret/error.rs:358]: "Invalid int ptr usage" doesn't really fit, as here it's not that the value itself is fundamentally wrong but that it gets used in the wrong way... maybe we should rename that variant?

`DanglingIntPtr` would be a possible name.

Comment (RalfJung) [reply src/librustc_middle/mir/interpret/error.rs:382]: @oli-obk like so?

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/71590 landed, so I rebased.
